delivery_line_qty_delivered/models/sale_order_line.py

```diff
@@ -0,0 +1,31 @@
+# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    qty_delivered_method = fields.Selection(selection_add=[("delivery", "Delivery")])
+
+    @api.depends("product_id", "is_delivery")
+    def _compute_qty_delivered_method(self):
+        super(SaleOrderLine, self)._compute_qty_delivered_method()
+        for line in self:
+            if line._is_delivery() and line.product_id.type == "service":
+                line.qty_delivered_method = "delivery"
+
+    @api.depends(
+        "order_id.delivery_state",
+    )
+    def _compute_qty_delivered(self):
+        super(SaleOrderLine, self)._compute_qty_delivered()
+        for line in self.filtered(lambda sol: sol.qty_delivered_method == "delivery"):
+            if (
+                line.product_id.invoice_policy == "delivery"
+                and line.order_id.delivery_state in ["done", "partially"]
+            ):
+                line.qty_delivered = line.product_uom_qty
```

Comment:
I would put that in a method `_get_delivery_qty_delivered()` as some may want a different value